### ISSUE TITLE
Assertion error when calculating loss

### ISSUE THREAD

**Issue by greg234234** [Curt-Park/rainbow-is-all-you-need#17]
**tl;dr**: an error occurs in [08.rainbow.ipynb](https://colab.research.google.com/github/Curt-Park/rainbow-is-all-you-need/blob/master/08.rainbow.ipynb) while training to 100,000 steps (sometimes < 20k steps, sometimes more), here's my copy of the [Colab](https://colab.research.google.com/drive/1dBJiv3wu1tScR-U84POjSNF0NVdJZiYj) 

Hi and thanks for sharing this wonderful learning resource, I really like your implementation! 

Testing with CartPole and LunarLander, inside colab, the network encounters `nan` inside `update_priorities(self, indices, priorities)` called by `loss = self.update_model()` inside the `train() function`.  For CartPole, this error comes just after 17500 steps.

![image](https://user-images.githubusercontent.com/15487280/62074089-02efc680-b210-11e9-955c-d8e2e1bacc69.png)

```
<ipython-input-20-92e9bbd87b9f> in train(self, num_frames, plotting_interval)
    222             # if training is ready
    223             if len(self.memory) >= self.batch_size:
--> 224                 loss = self.update_model()
    225                 losses.append(loss)
    226                 update_cnt += 1

<ipython-input-20-92e9bbd87b9f> in update_model(self)
    183         loss_for_prior = elementwise_loss.detach().cpu().numpy()
    184         new_priorities = loss_for_prior + self.prior_eps
--> 185         self.memory.update_priorities(indices, new_priorities)
    186 
    187         # NoisyNet: reset noise

<ipython-input-17-06b7969c0015> in update_priorities(self, indices, priorities)
     84 
     85         for idx, priority in zip(indices, priorities):
---> 86             assert priority > 0
     87             assert 0 <= idx < len(self)
     88 

AssertionError: 
```
You can trace this up the stack to the line
from `loss_for_prior = elementwise_loss.detach().cpu().numpy()`
to `elementwise_loss = self._compute_dqn_loss(samples, self.gamma)` 

So _compute_dqn_loss holds the clue because samples, nor self.gamma have nan values.

I've left a copy of this inside  [Colab](https://colab.research.google.com/drive/1dBJiv3wu1tScR-U84POjSNF0NVdJZiYj) which is a copy of 08.rainbow.ipynb configured for LunarLander-v2.  In other experiments I've played with the memory_size and set n_step to 1, but eventually the priorities variable receives a `nan`.  I'd love to hear any ideas why it might be happening 

For completeness, here's a copy of the current definition of that function _compute_dqn_loss:
```
    def _compute_dqn_loss(self, samples: Dict[str, np.ndarray], gamma: float) -> torch.Tensor:
        """Return categorical dqn loss."""
        device = self.device  # for shortening the following lines
        state = torch.FloatTensor(samples["obs"]).to(device)
        next_state = torch.FloatTensor(samples["next_obs"]).to(device)
        action = torch.LongTensor(samples["acts"]).to(device)
        reward = torch.FloatTensor(samples["rews"].reshape(-1, 1)).to(device)
        done = torch.FloatTensor(samples["done"].reshape(-1, 1)).to(device)
        
        # Categorical DQN algorithm
        delta_z = float(self.v_max - self.v_min) / (self.atom_size - 1)

        with torch.no_grad():
            # Double DQN
            next_action = self.dqn(next_state).argmax(1)
            next_dist = self.dqn_target.dist(next_state)
            next_dist = next_dist[range(self.batch_size), next_action]

            t_z = reward + (1 - done) * gamma * self.support
            t_z = t_z.clamp(min=self.v_min, max=self.v_max)
            b = (t_z - self.v_min) / delta_z
            l = b.floor().long()
            u = b.ceil().long()

            offset = (
                torch.linspace(
                    0, (batch_size - 1) * self.atom_size, self.batch_size
                ).long()
                .unsqueeze(1)
                .expand(self.batch_size, self.atom_size)
                .to(self.device)
            )

            proj_dist = torch.zeros(next_dist.size(), device=self.device)
            proj_dist.view(-1).index_add_(
                0, (l + offset).view(-1), (next_dist * (u.float() - b)).view(-1)
            )
            proj_dist.view(-1).index_add_(
                0, (u + offset).view(-1), (next_dist * (b - l.float())).view(-1)
            )

        dist = self.dqn.dist(state)
        log_p = torch.log(dist[range(self.batch_size), action])
        elementwise_loss = -(proj_dist * log_p).sum(1)

        return elementwise_loss

```
Thanks,
Greg

**Comment by greg234234** [comment]
Small update: I can confirm this occurs with all thee options for hardware acceleration

**Comment by Curt-Park** [comment]
Thanks for reporting. We are investigating this issue and will give you feedback soon.

**Comment by MrSyee** [comment]
Hi. @greg234234 !

Thank you for reporting! We check this part again.
As you said, `elementwise_loss` has a `nan` value in several cases. This is because `dist` has zero value, then it is same that softmax method outputs zero value.
(reference: https://github.com/intel/mkl-dnn/issues/30)
So calculating loss, it makes `-inf` from log(0), finally, we receive a `nan` value.

We resolve this problem by clipping `dist` on network class.
```
dist.clamp(min=1e-3)  # for avoiding nans
```

Thank you.

**Comment by greg234234** [comment]
Hi @MrSyee,
Thanks for your attention to this issue. I opened your new copy of 08.rainbow.ipynb and with "LunarLander-v2" it failed after a little over 100,000 steps. I also tried other environments.  Here is the snapshot, thank you! 

![image](https://user-images.githubusercontent.com/15487280/62573762-c69e1500-b864-11e9-8eb8-87b95e4ea888.png)

tensor([2.7745, 2.7312, 2.7209, 2.9932, 2.8715, 2.7502, 2.3976, 2.9276, 2.6649,
        2.3470, 2.6371, 2.7557, 2.9094,    nan, 2.5237, 2.6343, 2.9577, 2.7602,
        1.3124, 3.1003, 2.8163, 2.4867, 2.8150, 2.9254, 2.9196, 2.7499, 2.7402,
        2.6773, 3.0428, 2.9707, 3.0831, 2.9420], grad_fn=<NegBackward>)
tensor([2.7493, 2.7501, 2.6835, 2.9862, 2.9874, 2.7218, 2.3308, 2.9040, 2.7122,
        2.4148, 2.6528, 2.7502, 2.9114,    nan, 2.4950, 2.5737, 2.9077, 2.7332,
        -0.0000, 3.1718, 2.7998, 2.4378, 2.7984, 2.9034, 2.8745, -0.0000, 2.7189,
        2.6672, 3.0187, 2.9696, 3.0807, 2.9211], grad_fn=<NegBackward>)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-13-4326fee8443c> in <module>()
----> 1 agent.train(num_frames)

2 frames
<ipython-input-7-de3ab62f8e40> in train(self, num_frames, plotting_interval)
    222             # if training is ready
    223             if len(self.memory) >= self.batch_size:
--> 224                 loss = self.update_model()
    225                 losses.append(loss)
    226                 update_cnt += 1

<ipython-input-7-de3ab62f8e40> in update_model(self)
    183         loss_for_prior = elementwise_loss.detach().cpu().numpy()
    184         new_priorities = loss_for_prior + self.prior_eps
--> 185         self.memory.update_priorities(indices, new_priorities)
    186 
    187         # NoisyNet: reset noise

<ipython-input-4-06b7969c0015> in update_priorities(self, indices, priorities)
     84 
     85         for idx, priority in zip(indices, priorities):
---> 86             assert priority > 0
     87             assert 0 <= idx < len(self)
     88 

AssertionError:

**Comment by MrSyee** [comment]
@greg234234

Oh, It's a mistake. It needs the left value.
`dist = dist.clamp(min=1e-3)  # for avoiding nans`

I push again after correcting.

Thank you for your review. :)

**Comment by greg234234** [comment]
@MrSyee  it's a pleasure, thanks for the quick fix.  A little embarrassed that I didn't notice it either haha.   You guys are awesome, I'd like very much to help more.

**Comment by MrSyee** [comment]
@greg234234 Thank you for your compliment. :)

@Curt-Park Please review PR.
